core/async/src/test_loop.rs

```diff
@@ -0,0 +1,372 @@
+        }
+        for handler in &mut self.handlers {
+            handler.init(self.sender.clone());
+        }
+    }
+
+    /// Runs the test loop for the given duration. This function may be called
+    /// multiple times, but further test handlers may not be registered after
+    /// the first call.
+    pub fn run(&mut self, duration: Duration) {
+        self.maybe_initialize_handlers();
+        let deadline = self.current_time + duration;
+        'outer: loop {
+            // Push events we have just received into the heap.
+            while let Ok(event) = self.pending_events.try_recv() {
+                self.events.push(EventInHeap {
+                    due: self.current_time + event.delay,
+                    event: event.event,
+                    id: self.next_event_index,
+                });
+                self.next_event_index += 1;
+            }
+            // Don't execute any more events if we have reached the deadline.
+            match self.events.peek() {
+                Some(event) => {
+                    if event.due >= deadline {
+                        break;
+                    }
+                }
+                None => break,
+            }
+            // Find the next event, log a line about it, and execute.
+            let event = self.events.pop().unwrap();
+            let json_printout = serde_json::to_string(&EventStartLogOutput {
+                current_index: event.id,
+                total_events: self.next_event_index,
+                current_event: format!("{:?}", event.event),
+                current_time_ms: event.due.as_millis() as u64,
+            })
+            .unwrap();
+            info!(target: "test_loop", "TEST_LOOP_EVENT_START {}", json_printout);
+            self.current_time = event.due;
+
+            let mut current_event = event.event;
+            for handler in &mut self.handlers {
+                if let Some(event) = handler.handle(current_event, &mut self.data) {
+                    current_event = event;
+                } else {
+                    continue 'outer;
+                }
+            }
+            panic!("Unhandled event: {:?}", current_event);
+        }
+    }
+}
+
+/// A convenient trait to TryInto, or else return the original object. It's useful
+/// for implementing event handlers.
+pub trait TryIntoOrSelf<R>: Sized {
+    fn try_into_or_self(self) -> Result<R, Self>;
+}
+
+impl<R, T: TryInto<R, Error = derive_more::TryIntoError<T>>> TryIntoOrSelf<R> for T {
+    fn try_into_or_self(self) -> Result<R, Self> {
+        self.try_into().map_err(|err| err.input)
+    }
+}
+
+/// An event handler that puts the event into a vector in the Data, as long as
+/// the Data contains a Vec<CapturedEvent>.
+///
+/// This is used on output events so that after the test loop finishes running
+/// we can assert on those events.
+pub struct CaptureEvents<CapturedEvent> {
+    _marker: std::marker::PhantomData<CapturedEvent>, // just for the compiler
```

Comment:
Do you want the users to create this struct only using `fn new()`?
`just for the compiler` isn't very descriptive.

Reply:
Also, ideally this can be implemented without PhantomData by doing something like

struct CaptureEvents;

pub fn capture_events<Event>() -> impl EventHandler<Event> {
  return CaptureEvents;
}

but LoopEventHandler also has a Data parameter. This prompts me to think about what if we create another handler trait, like

```
trait SpecificLoopEventHandler<Data, Event> {
  fn handle(... same thing);
}
```

and have that implement LoopEventHandler by doing an automatic conversion:

```
impl <
  Data,
  Event,
  A: SpecificLoopEventHandler<Data, Event>,
  OuterData: AsMut<Data>,
  OuterEvent: TryIntoOrSelf<Event> + From<Event>
> LoopEventHandler<OuterData, OuterEvent> for A {
  ...
}
```

but that also doesn't work because then we get an error for Data and Event that they are unconstrained type parameters. It seems like always a giant pain to implement a trait for everything that implements another trait.

Would be nice to make this work though because then it's easier to write a handler without having to use AsMut, TryIntoOrSelf magic. I'll play around with this more.